src/Orleans/Core/InvocationMethodInfoMap.cs

```diff
@@ -1,52 +1,45 @@
-            new ConcurrentDictionary<Type, ConcurrentDictionary<int, IReadOnlyDictionary<int, MethodInfo>>>();
-
         /// <summary>
-        /// Returns the <see cref="MethodInfo"/> for the specified implementation and invocation request.
+        /// The map from implementation types to interface ids to invoker.
         /// </summary>
-        /// <param name="implementationType">The type of the implementation.</param>
-        /// <param name="request">The invocation request.</param>
-        /// <returns>The <see cref="MethodInfo"/> for the specified implementation and invocation request.</returns>
-        public MethodInfo GetMethodInfo(Type implementationType, InvokeMethodRequest request)
+        private readonly ConcurrentDictionary<Type, ConcurrentDictionary<int, InterceptedMethodInvoker>> invokers =
+            new ConcurrentDictionary<Type, ConcurrentDictionary<int, InterceptedMethodInvoker>>();
+
+        public InterceptedMethodInvoker GetInterceptedMethodInvoker(Type implementationType, int interfaceId, IGrainMethodInvoker invoker)
         {
-            var implementation = this.implementations.GetOrAdd(
+            var implementation = invokers.GetOrAdd(
                 implementationType,
-                _ => new ConcurrentDictionary<int, IReadOnlyDictionary<int, MethodInfo>>());
+                _ => new ConcurrentDictionary<int, InterceptedMethodInvoker>());
 
-            IReadOnlyDictionary<int, MethodInfo> interfaceMap;
-            if (!implementation.TryGetValue(request.InterfaceId, out interfaceMap))
+            InterceptedMethodInvoker interceptedMethodInvoker;
+            if (implementation.TryGetValue(interfaceId, out interceptedMethodInvoker))
             {
-                // Get the interface mapping for the current implementation.
-                var interfaces = GrainInterfaceData.GetRemoteInterfaces(implementationType);
-                Type interfaceType;
-                if (!interfaces.TryGetValue(request.InterfaceId, out interfaceType))
-                {
-                    // The specified type does not implement the provided interface.
-                    return null;
-                }
+                return interceptedMethodInvoker;
+            }
 
-                // Create a mapping between the interface and the implementation.
-                interfaceMap = implementation.GetOrAdd(
-                    request.InterfaceId,
-                    _ => MapInterfaceToImplementation(interfaceType, implementationType));
+            // Get the interface mapping for the current implementation.
+            var interfaces = GrainInterfaceData.GetRemoteInterfaces(implementationType);
+            Type interfaceType;
+            if (!interfaces.TryGetValue(interfaceId, out interfaceType))
+            {
+                // The specified type does not implement the provided interface.
+                return null;
```

Comment:
If we return null here, https://github.com/dotnet/orleans/pull/963/files?diff=split#diff-6be22c3f41e935d133324f639e29b32aR371 will throw a NRE, right? Should we explicitly throw here instead?